File: kdtree/kdtree.c
#include <stdlib.h>
#include <string.h>
#include <math.h>
#include "kdtree.h"

#ifdef USEOMP
/* use OpenMP */
#include <omp.h>
#endif

/* 
kdtree_build build k-d tree for fast nearest neighbors.  This routine
is meant to be called as a dll from Matlab--that's why no structures.
*/

long int kdtree_build_idl( int argc, void *argv[]) {
  /* expects all arguments passed by reference */
  return(kdtree_build((double *)argv[0], /* *X */
		      *(unsigned long int *)argv[1], /* Nx */
		      *(unsigned long int *)argv[2], /* Nc */
		      *(int *)argv[3], /* flags */
		      (unsigned long int *)argv[4], /* *root */
		      (unsigned short int *)argv[5], /* *c */
		      (unsigned long int *)argv[6], /* *parent */
		      (unsigned long int *)argv[7], /* *left */
		      (unsigned long int *)argv[8])); /* *right */
}
int kdtree_build(const double *X, 
		 const unsigned long int Nx,
		 const unsigned long int Nc,
		 const int flags,
		 unsigned long int *root,
		 unsigned short int *c,
		 unsigned long int *parent,
		 unsigned long int *left,
		 unsigned long int *right) {

  const short int XmajorFlag = (flags & 1);
  
  inline unsigned long int ss2(const unsigned long int Nrows, 
				      const unsigned long int Ncols,
				      const unsigned long int i,
				      const unsigned long int j) { 
    /* 1-based 2-D subscript to 0-based 1-D subscript */
    /* XmajorFlag - 0 for ColumnMajor, 1 for RowMajor, 
       (controls majority of all input and output matrices) */
    return(XmajorFlag?(Ncols*(i-1)+(j-1)):(Nrows*(j-1)+(i-1)));
  }

  unsigned long int *new_range(const unsigned long int i1, const unsigned long int i2, unsigned long int *NI) {
    unsigned long int i,*I = malloc((i2-i1+1)*sizeof(unsigned long int));
    for (i=i1; i <= i2; i++) {
      I[i-i1] = i;
    }
    if (NI) {
      *NI = (i2-i1+1);
    }
    return(I);
  }

  unsigned long int *new_reindex(const unsigned long int *I, const unsigned long int *i1, unsigned long int Ni1) {
    /* return I(i1), where i1 is a set of Ni1 1-based indexes */
    unsigned long int i,*I1 = malloc(Ni1*sizeof(unsigned long int));
    for (i=0; i < Ni1; i++) {
      I1[i] = I[i1[i]-1];
    }
    return(I1);
  }

  unsigned long int kdtree_split(const unsigned long int *I, const unsigned long int NI,const short int c_in,const unsigned long int iparent) {

    int XIcompare(const  void *i1, const void *i2) {
      double x1 = X[ss2(Nx,Nc,*((unsigned long int*)i1),c_in)];
      double x2 = X[ss2(Nx,Nc,*((unsigned long int*)i2),c_in)];
      if (x1==x2) {
	return(0);
      } else if (x1 < x2) {
	return(-1);
      } else {
	return(+1);
      }
    }

    unsigned long int *Isorted = malloc(NI*sizeof(unsigned long int));

    /* get Isorted that sorts X(Isorted,c) */
    memcpy(Isorted,I,NI*sizeof(unsigned long int));
    qsort(Isorted,NI,sizeof(unsigned long int),&XIcompare);

    unsigned long int jmid = ceil((double)NI/2.0);
    unsigned long int leaf_index = Isorted[jmid-1];

    if (iparent==0) {
      *root = leaf_index;
    }
    c[leaf_index-1] = c_in;
    parent[leaf_index-1] = iparent;

    /* prepare to call split */
    short int c_out = 1+(c_in % Nc);

    if (jmid>1) {
      unsigned long int NI1,*i1 = new_range(1,jmid-1,&NI1);
      unsigned long int *I1 = new_reindex(Isorted,i1,NI1);
      left[leaf_index-1] = kdtree_split(I1,NI1,c_out,leaf_index);
      free(i1);
      free(I1);
    }

    if (jmid < NI) {
      unsigned long int NI2,*i2 = new_range(jmid+1,NI,&NI2);
      unsigned long int *I2 = new_reindex(Isorted,i2,NI2);
      right[leaf_index-1] = kdtree_split(I2,NI2,c_out,leaf_index);
      free(i2);
      free(I2);
    }
    free(Isorted);
    return(leaf_index);
  }
  
  *root = 0;
  /* assume c, parent, left, right are allocated already */
  /* set to zero */
  memset(c,0,Nx*sizeof(*c));
  memset(parent,0,Nx*sizeof(*parent));
  memset(left,0,Nx*sizeof(*left));
  memset(right,0,Nx*sizeof(*right));

  unsigned long int NI,*I = new_range(1,Nx,&NI);

  kdtree_split(I,NI,1,0);

  free(I);

  return(1); /* return value presently not used */
}

int kdtree_kNN_range(const unsigned long int i1,
		     const unsigned long int i2,
		     const double *X, 
		     const unsigned long int Nx,
		     const unsigned long int Nc,
		     const int flags,
		     const unsigned long int root,
		     const unsigned short int *c,
		     const unsigned long int *parent,
		     const unsigned long int *left,
		     const unsigned long int *right,
		     const double *X0,
		     const unsigned long int NX0,
		     const unsigned long int k,
		     const double *DistScale,
		     unsigned long int *index,
		     double *R2) {

  /* i1, i2 1-based indices for which X0's to do in this call (helps parallilization) */

  const short int XmajorFlag = (flags & 1);
  const double double_inf = +1./0.0;  /* positive infinity */

    inline unsigned long int ss2(const unsigned long int Nrows, 
				 const unsigned long int Ncols,
				 const unsigned long int i,
				 const unsigned long int j) { 
      /* 1-based 2-D subscript to 0-based 1-D subscript */
      /* XmajorFlag - 0 for ColumnMajor, 1 for RowMajor, 
	 (controls majority of all input and output matrices) */
      return(XmajorFlag?(Ncols*(i-1)+(j-1)):(Nrows*(j-1)+(i-1)));
    }
    
    double dist_fun1(unsigned long int iX, unsigned long int iX0,unsigned long int j) {
      /* signed, scaled distance between X(iX,:) and X0(iX0,:) in the j coordinate */
      double dist = (X[ss2(Nx,Nc,iX,j)]-X0[ss2(NX0,Nc,iX0,j)]);
      if (DistScale) {
	dist *= DistScale[j-1];
      }
      return(dist);
    }
    
    double dist_fun_sumsq(unsigned long int iX, unsigned long int iX0) {
      /* sum squared, scaled distance between X(iX,:) and X0(iX0,:) */
      unsigned long int j;
      double dist = 0;
      for (j=1; j <= Nc; j++) {
	double tmp = dist_fun1(iX,iX0,j);
	dist += tmp*tmp;
      }
      return(dist);
    }

    double dist_fun_maxsq(unsigned long int iX, unsigned long int iX0) {
      /* maximum squared, scaled distance between X(iX,:) and X0(iX0,:) */
      unsigned long int j;
      double dist = 0;
      for (j=1; j <= Nc; j++) {
	double tmp = dist_fun1(iX,iX0,j);
	tmp *= tmp; /* tmp^2 */
	if (tmp>dist) {
	  dist = tmp;
	}
      }
      return(dist);
    }

    /* flags & 2 == 0 for pythagorean distance, 2 for max absolute difference */
  double (*dist_fun)(unsigned long int, unsigned long int);
  if (flags & 2) {
    dist_fun = &dist_fun_maxsq;
  } else {
    dist_fun = &dist_fun_sumsq;
  }

    
    void check_node(unsigned long int i,
		    unsigned long int iX0,
		    short int check_parent,
		    short int check_left,
		    short int check_right) {
      /* check node and recursively check its
	 parent, left child, and/or right child
	 i is an index into the tree variables (X, parent, right, left, c, etc)
	 iX0 is an index into the center points matrix (X0)
      */
      if (i<=0) {
	return;
      }
      
      double dxc = dist_fun1(i,iX0,c[i-1]); /* signed distance along c */
      double dxc2 = dxc*dxc; /* squared distance along c */
      unsigned long int ssR2; /* subscript for iR2 */
      if (dxc2 <= R2[ssR2 = ss2(NX0,k,iX0,k)]) { /* check against farthest nearest neighbor */
	/* this node and its children could be nearest neighbors */
	double r2i = (*dist_fun)(i,iX0); /* full Nc-d squared distance */
	if (r2i <= R2[ssR2]) { /* this is an active nearest neighbor */
	  unsigned long int iR2 = k,ssR2m1; /* ssR2m1 - subscript for iR2-1 */
	  /* walk inward from farthest nearest neighbor,
	     copying distances larger than r2i outward,
	     until we find a spot for this node
	  */
	  while ((iR2 > 1) && (r2i <= R2[ssR2m1 = ss2(NX0,k,iX0,iR2-1)])) {
	    R2[ssR2] = R2[ssR2m1];
	    index[ssR2] = index[ssR2m1];
	    iR2--;
	    ssR2 = ssR2m1;
	  } /* end while */
	  /* now we have a spot for this node, store it */
	  R2[ssR2] = r2i;
	  index[ssR2] = i;
	} /* end if r2i */
	
	/* ok, check both children if allowed */
	if (check_right) {
	  check_node(right[i-1],iX0,0,1,1);
	}
	if (check_left) {
	  check_node(left[i-1],iX0,0,1,1);
	}
      } else {
	/* this node is too far away, but its kids might be closer */
	if (dxc<0) { /* only kids to the right can have smaller dxc2, and be in range */
	  if (check_right) {
	    check_node(right[i-1],iX0,0,1,1);
	  }
	} else { /* only kids to the left can have smaller dxc2, and be in range */
	  if (check_left) {
	    check_node(left[i-1],iX0,0,1,1);
	  }
	} /* end if dxc<0 */
      } /*end if dxc2 */
      if (check_parent && (parent[i-1]>0)) {
	/* check parent, but don't recheck this node */
	short int isright = (right[parent[i-1]-1] == i); /* is this a right child of parent? */
	check_node(parent[i-1],iX0,1,isright,!isright);
      }
      
    }

    unsigned long int iX0;
    for (iX0 = i1; iX0 <= i2; iX0++) {
      unsigned long int i,ilast;
      /* prepare distance array w/ infs */
      for (i=1; i <= k; i++) {
      	R2[ss2(NX0,k,iX0,i)] = double_inf;
      }
      /* descend to closest leaf node */
      i = root;
      while (i>0) {
      	ilast = i;
      	if (X0[ss2(NX0,Nc,iX0,c[i-1])] > X[ss2(Nx,Nc,i,c[i-1])]) {
      	  i = right[i-1];
      	} else {
      	  i = left[i-1];
      	}
      } /* end while */
      /* ilast now points to nearest leaf node (may have one branch in away direction */
      check_node(ilast,iX0,1,1,1); /* check node, its parent and any children */
    } /* end of for loop */
    
  return(1); /* return value presently not used */
}

	       
long int kdtree_kNN_idl( int argc, void *argv[]) {
  /* expects all arguments passed by reference */
  return(kdtree_kNN(argv[0], /* *X */
		    *(unsigned long int*)argv[1], /* Nx */
		    *(unsigned long int*)argv[2], /* Nc */
		    *(int*)argv[3], /* flags */
		    *(unsigned long int*)argv[4], /* root */
		    (unsigned short int*)argv[5], /* *c */
		    (unsigned long int*)argv[6], /* *parent */
		    (unsigned long int*)argv[7], /* *left */
		    (unsigned long int*)argv[8], /* *right */
		    (double*)argv[9], /* *X0 */
		    *(unsigned long int*)argv[10], /* NX0 */
		    *(unsigned long int*)argv[11], /* k */
		    (double*)argv[12], /* *DistScale */
		    (unsigned long int*)argv[13], /* *index */
		    (double*)argv[14])); /* *R2 */
}

int kdtree_kNN(const double *X, 
	       const unsigned long int Nx,
	       const unsigned long int Nc,
	       const int flags,
	       const unsigned long int root,
	       const unsigned short int *c,
	       const unsigned long int *parent,
	       const unsigned long int *left,
	       const unsigned long int *right,
	       const double *X0,
	       const unsigned long int NX0,
	       const unsigned long int k,
	       const double *DistScale,
	       unsigned long int *index,
	       double *R2) {
  short int freeR2=0;
  int result = 0;

  if (!R2) {
    /* allocate distance array */
    R2 = malloc(NX0*k*sizeof(double));
    freeR2 = 1; /* will need to free later */
  }

#ifdef USEOMP
#pragma omp parallel
  { /* start parallel */
    int num_threads = 1,tid=0;
    /* i1, i2 1-based indices for which X0's to do in the sub-call */
    /* Ntsub Number of time steps per sub */
    unsigned long int i1,i2,Ntsub;

    tid = omp_get_thread_num();
    num_threads = omp_get_num_threads();
    Ntsub = NX0/num_threads;

    i1 = 1+tid*Ntsub;
    if (tid == num_threads-1) {
      i2 = NX0;
    } else {
      i2 = i1+Ntsub-1;
    }
    
    kdtree_kNN_range(i1,i2,X,Nx,Nc,flags,root,c,parent,left,right,X0,NX0,k,DistScale,index,R2);
    result = 1;
  } /* end parallel */
#else
  result = kdtree_kNN_range(1,NX0,X,Nx,Nc,flags,root,c,parent,left,right,X0,NX0,k,DistScale,index,R2);
#endif

  /* free memory */
  if (freeR2) {
    free(R2);
  }

  return(result);
};

File: kdtree/matlab/kdtree.m
function varargout = kdtree(what,varargin)
% tree = kdtree('build',X) % build kdtree
%
% [index,R2,XNN] = kdtree('kNN',X,tree,X0,k,DistScale) % retrieve index and (optionally location XNN) of k points nearest to X0
% DistScale (size(X0))is optional - defines weighting for each dimension
%
% [index,R2,XNN] = kdtree('multikNN',X,tree,X0,k,DistScale) % retrieve index and (optionally location XNN) of k points nearest to X0
% ... = kdtree('multikNN',...,'max') % use maximum squared difference
%           rather than sum squared difference for distance measure
%           (default is 'sum')
% DistScale (length size(X0,2))is optional - defines weighting for each dimension
% optionally handles multiple X0, where each is a row vector
% size(index) and size(R2) are [k size(X0,1)]
% size(XNN) is [k size(X0,1) size(X0,2)]
% requires kdtree.dll (or .so) and kdtree.h in the matlab path
% R2 is squared distance

varargout = cell(1,nargout);

switch(lower(what)),
    case {'nolibbuild'},
        [varargout{:}] = kdtree_build(varargin{1});
    case {'libbuild','build'},
        [varargout{:}] = kdtree_libbuild(varargin{1});
    case {'knn'},
        [varargout{:}] = kdtree_kNN(varargin{:});
    case {'multiknn'},
        [varargout{:}] = kdtree_multikNN(varargin{:});
    otherwise
        error('Unknown what "%s"',what);
end

end

function tree = kdtree_build(X)

% size variables
[Nx,Nc] = size(X);
if (Nx==0),
    error('Cannot create empty kdtree');
end

% save memory, use smallest necessary int type for index records
itype = get_int_type(Nx);
ctype = get_int_type(Nc);

tree.root = itype(0);
tree.c = ctype(zeros(Nx,1));
tree.parent = itype(zeros(Nx,1));
tree.left = itype(zeros(Nx,1));
tree.right = itype(zeros(Nx,1));

[foo,tree] = kdtree_split(X,tree,itype(1:size(X,1))',ctype(1),itype(0));

end % end of function


function tree = kdtree_libbuild(X)
% size variables
[Nx,Nc] = size(X);
if Nx==0,
    error('Cannot create empty kdtree');
end
flags = 0; % Matlab is column major (0)
kdtree_loadlib;

if isoctave,
    [tree.root,tree.c,tree.parent,tree.left,tree.right] = kdtree_build_oct(X,Nx,Nc,flags);
else
    
    % int kdtree_build(const double *X,
    % 		 const unsigned long int Nx,
    % 		 const unsigned long int Nc,
    % 		 const int flags,
    % 		 unsigned long int *root,
    % 		 unsigned short int *c,
    % 		 unsigned long int *parent,
    % 		 unsigned long int *left,
    % 		 unsigned long int *right);
    
    
    
    
    rootPtr = libpointer('ulongPtr',0);
    cPtr = libpointer('uint16Ptr',zeros(Nx,1)); % assumes ushort = uint16, should be true
    parentPtr = libpointer('ulongPtr',zeros(Nx,1));
    leftPtr = libpointer('ulongPtr',zeros(Nx,1));
    rightPtr = libpointer('ulongPtr',zeros(Nx,1));
    
    calllib('kdtree','kdtree_build',X,Nx,Nc,flags,rootPtr,cPtr,parentPtr,leftPtr,rightPtr);
    
    tree.root = rootPtr.value;
    tree.c = cPtr.value;
    tree.parent = parentPtr.value;
    tree.left = leftPtr.value;
    tree.right = rightPtr.value;
end

end % end of function

function [leaf_index,tree] = kdtree_split(X,tree,I,c,iparent)

[Nx,Nc] = size(X);
NI = size(I,1);

[xc,isort] = sort(X(I,c));
I = I(isort);

jmid = ceil(NI/2);
leaf_index = I(jmid);

if iparent==0,
    tree.root = leaf_index;
end

tree.c(leaf_index) = c;
tree.parent(leaf_index) = iparent;

% prepare to call split
c = 1+rem(c,Nc);
i1 = (1:(jmid-1));
i2 = ((jmid+1):NI);

if ~isempty(i1),
    [leaf,tree] = kdtree_split(X,tree,I(i1),c,leaf_index);
    tree.left(leaf_index) = leaf;
end
if ~isempty(i2),
    [leaf,tree] = kdtree_split(X,tree,I(i2),c,leaf_index);
    tree.right(leaf_index) = leaf;
end
end % end of function


function varargout = kdtree_kNN(X,tree,X0,k,DistScale)

if nargin < 5,
    DistScale = ones(1,size(X0,2));
    dist_fun = @(x,y) sum((x(:)-y(:)).^2);
else
    dist_fun = @(x,y) sum((DistScale(:).*(x(:)-y(:))).^2);
end

varargout = cell(1,nargout);

persistent warned
if isoctave,
    if isempty(warned) || ~warned,
        warning('***kdtree_kNN does not work in octave because it uses nested functions. Using multikNN instead');
        warned = true;
    end
    [varargout{:}] = kdtree_multikNN(X,tree,X0,k,DistScale);
    return
end


% variables shared with subfunctions: X, tree, Nc, index, XNN, R2, dist_fun, DistScale
Nc = size(X0,2);
index = zeros(k,1,class(tree.parent(1)));
R2 = inf(k,1);
if nargout >= 3,
    XNN = nan(k,Nc);
else
    XNN = [];
end

% first we descend to closest point
i = tree.root;
while i>0,
    ilast = i;
    if X0(tree.c(i))>X(i,tree.c(i)),
        i = tree.right(i);
    else
        i = tree.left(i);
    end
end

% now ilast points to the closest leaf to X0

% check parent nodes for closer children
check_node(ilast,true,true,true);

varargout{1} = index;
if nargout >= 2,
    varargout{2} = R2;
end
if nargout >= 3,
    varargout{3} = XNN;
end

    function check_node(i,check_parent,check_left,check_right)
        if i <= 0, % null node
            return;
        end
        dxc = DistScale(tree.c(i))*(X(i,tree.c(i))-X0(tree.c(i)));
        dxc2 = dxc^2;
        if dxc2 <= R2(end), % this node and its children could be nearest neighbors
            r2i = dist_fun(X(i,:),X0);
            if r2i <= R2(end), % this is an active nearest neighbor
                iR2 = length(R2);
                while (iR2 > 1) && (r2i <= R2(iR2-1)),
                    R2(iR2) = R2(iR2-1);
                    index(iR2) = index(iR2-1);
                    if ~isempty(XNN),
                        XNN(iR2,:) = XNN(iR2-1,:);
                    end
                    iR2 = iR2 - 1;
                end
                R2(iR2) = r2i;
                index(iR2) = i;
                if ~isempty(XNN),
                    XNN(iR2,:) = X(i,:);
                end
            end
            if check_right,
                check_node(tree.right(i),false,true,true);
            end
            if check_left,
                check_node(tree.left(i),false,true,true);
            end
        else
            % this node is too far away, but its kids might be
            if dxc<0, % only kids to the right can have smaller dxc2
                if check_right,
                    check_node(tree.right(i),false,true,true);
                end
            else % only kids to the left can have smaller dxc2
                if check_left,
                    check_node(tree.left(i),false,true,true);
                end
            end
        end
        if check_parent && (tree.parent(i)>0),
            % check parent, but don't recheck this node
            isright = (tree.right(tree.parent(i))==i);
            check_node(tree.parent(i),true,isright,~isright);
        end
    end

end

function [index,R2,XNN] = kdtree_multikNN(X,tree,X0,k,varargin)

DistScale = [];
how = 'sum'; % how to combine differences to compute distance
if length(varargin)==2,
    DistScale = varargin{1};
    how = varargin{2};
elseif length(varargin)==1,
    if ischar(varargin{1}),
        how = varargin{1};
    else
        DistScale = varargin{1};
    end
end

switch(lower(how)),
    case 'max',
        how_flag = 2;
    case 'sum',
        how_flag = 0;
    otherwise
        error('how="%s" not understood',how);
end

[Nx,Nc] = size(X);
NX0 = size(X0,1);
flags = 0; % Matlab is column major (0)
flags = flags + how_flag; % add flag for how to compbine differences into dist
kdtree_loadlib;

if isoctave,
    [index,R2] = kdtree_multikNN_oct(X,Nx,Nc,flags,tree.root,tree.c,tree.parent,tree.left,tree.right,X0,NX0,k,DistScale);
    index = index';
    R2 = R2';
else % end of octave specific code
    
    if isempty(DistScale),
        DistScale = libpointer('doublePtr'); % null
    end
    
    indexPtr = libpointer('ulongPtr',zeros(NX0,k));
    if nargout >= 2,
        R2Ptr = libpointer('doublePtr',zeros(NX0,k));
    else
        R2Ptr = libpointer('doublePtr'); % null
    end
    
    %int32 kdtree_kNN(doublePtr, ulong, ulong, int32, ulong, uint16Ptr, ulongPtr, ulongPtr, ulongPtr, doublePtr, ulong, ulong, doublePtr, ulongPtr, doublePtr)
    

    calllib('kdtree','kdtree_kNN',X,Nx,Nc,flags,...
        tree.root,tree.c,tree.parent,tree.left,tree.right,...
        X0,NX0,k,DistScale,indexPtr,R2Ptr);
    
    index = get(indexPtr,'val')';
    if nargout >= 2,
        R2 = get(R2Ptr,'val')';
    end
    
end % end of matlab-specific code

if nargout >= 3, % populate XNN output
    XNN = nan([k size(X0)]);
    for c = 1:Nc,
        XNN(:,:,c) = reshape(X(index,c),size(index));
    end
end

end % end of function


function itype = get_int_type(N)

Ibits = ceil(log(1+N)/log(2));

if Ibits <= 8,
    itype = @uint8;
elseif Ibits <= 16,
    itype = @uint16;
elseif Ibits <= 32,
    itype = @uint32;
elseif Ibits <= 64,
    itype = @uint64;
end

end

function kdtree_loadlib()
if isoctave,
    if ~exist('kdtree_multikNN_oct'), % load
        octfile = file_in_loadpath('kdtree_oct.oct');
        if isempty(octfile),
            error('unable to locate kdtree_oct.oct');
        end
        autoload('kdtree_multikNN_oct',octfile);
        autoload('kdtree_build_oct',octfile);
    end
else
    if ~libisloaded('kdtree'),
        
        if ~exist('kdtree.h','file'),
            error('Unable to locate kdtree.h');
        end
        
        if ispc,
            libname = 'kdtree.dll';
        else
            libname = 'kdtree.so';
        end
        
        loadlibrary(libname,'kdtree.h','alias','kdtree');
        
    end
end
end
